src/validation/manifest.js

```diff
@@ -51,13 +51,112 @@ const validators = immutable.fromJS({
   ScalarTypeDefinition: (value, ctx) =>
     validators.get(ctx.getIn(['type', 'name', 'value']))(value, ctx),
 
-  UnionTypeDefinition: (value, ctx) =>
-    ctx
-      .getIn(['type', 'types'])
-      .reduce(
-        (errors, type) => errors.concat(validateValue(value, ctx.set('type', type))),
-        List(),
-      ),
+  UnionTypeDefinition: (value, ctx) => {
+    // Concat path for type deduction
+    const path = ctx.get('path').toJS().reduce((prev, current, index) => {
+      if (index === 0) {
+        return current;
+      } else if (typeof current === 'number') {
+        return `${prev}[$]`
+      } else {
+        return `${prev}.${current}`
+      }
+    }, '');
+
+    // Deduce type of the union
+    let unionType;
+
+    switch (path) {
+      case "mutations": {
+        if (value.get('file')) {
+          unionType = "Manifest";
+        } else {
+          unionType = "MutationsManifest";
+        }
+        break;
+      }
+      case "mutations.resolvers": {
+        if(!value.get('kind')){
+          return immutable.fromJS([
+            {
+              path: [...ctx.get('path'), 'kind'],
+              message: `No value provided`,
+            },
+          ])
+        } else if (value.get('kind') === "javascript") {
+          unionType = "JavascriptResolvers"
+        } else {
+          return immutable.fromJS([
+            {
+              path: [...ctx.get('path'), 'kind'],
```

Comment:
I can see how we'd want a more specific path when validating union type values, but I disagree with how it's done here. This `UnionTypeDefinition` dispatcher is supposed to be generic. Data sources and templates were already being validated before, there is no need to hard-code anything about them here, I think.

We should instead have two specific additional type validators: One for `ExternalMutations` and one for `InlineMutations`. (Ok, and one for the resolvers.)

Reply:
The reason we went about writing this is because the original UnionTypeDefinition wasn't actually functional for Unions of multiple types (code [here](https://github.com/graphprotocol/graph-cli/blob/a4b62fab8138e298455e96ad0acba95a5eecfd4a/src/validation/manifest.js#L56)). It would iterate through all union types, and aggregate the errors from each one. Instead of doing this, we instead want to have some way of determine what union type we need to use. This is done by looking at the `kind` property in some cases (datasource types for example), and also by looking for the presence of a special determining field like the `file` field in the `ExternalMutations` type.

It is for the above reasons that I think we need to have programmatic type deducers. Would you agree? Another route is to iterate through all union types, and if one succeeds, use that as the resulting type and disregard all other errors. This would lead to some weird errors being returned if none succeed, as all possible errors form all possible types would be returned.